src/org/jgroups/raft/filelog/FilePositionCache.java

```diff
@@ -11,77 +12,222 @@
 public class FilePositionCache {
 
    public static final int TOO_OLD = -1;
-   public static final int NO_CAPACITY = -2;
-   public static final int OK = 1;
-   private static final int SIZE_INCREMENT = 128;
 
-   private final long[] position;
+   private static final int PAGE_CAPACITY = 1 << 10;
+   private static final int PAGE_SHIFT = Integer.numberOfTrailingZeros(PAGE_CAPACITY);
+   private static final int PAGE_MASK = PAGE_CAPACITY - 1;
+   private static final int EMPTY = -1;
+
+   {
+      assert Integer.bitCount(PAGE_CAPACITY) == 1 : "PAGE_CAPACITY MUST BE A POWER OF TWO";
+   }
+
+   private static final class PositionPage {
+
+      private final long[] positions;
+      private int lastNotEmptyOffset;
+
+      public PositionPage() {
+         positions = new long[PAGE_CAPACITY];
+         Arrays.fill(positions, EMPTY);
+         lastNotEmptyOffset = -1;
+      }
+
+      public boolean clear() {
+         return clearFrom(0);
+      }
+
+      public boolean clearFrom(final int pageOffset) {
+         if (pageOffset > lastNotEmptyOffset) {
+            return false;
+         }
+         if (pageOffset < 0) {
+            throw new IllegalArgumentException("pageOffset must be greater then 0");
+         }
+         if (pageOffset >= PAGE_CAPACITY) {
+            throw new IllegalArgumentException("pageOffset must be less then page capacity");
+         }
+         assert lastNotEmptyOffset >= 0;
```

Comment:
unnecessary in my opinion. It is impossible to fail since we check `pageOffset > lastNotEmptyOffset` and `pageOffset >= 0`